### PR TITLE
test(e2e): add stream tcp proxy with SNI test 

### PR DESCRIPTION
**Pull request by Juneezee**
### Type of change:

- [x] New test case

### What this PR does / why we need it:

Closes: https://github.com/apache/apisix-ingress-controller/issues/1438

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by Juneezee** [comment]
The test passes if I modify the code to create `TLSRoute`

```go
	ginkgo.It("stream tcp proxy with SNI", func() {
		// create secrets
		host := "a.test.com"
		secret := "server-secret"
		serverCert, serverKey := generateCert(ginkgo.GinkgoT(), []string{host})
		err := s.NewSecret(secret, serverCert.String(), serverKey.String())
		assert.Nil(ginkgo.GinkgoT(), err, "create server cert secret error")

		// create ApisixTls resource
		err = s.NewApisixTls("tls-server", host, secret)
		assert.Nil(ginkgo.GinkgoT(), err, "create ApisixTls error")

		// check ssl in APISIX
		assert.Nil(ginkgo.GinkgoT(), s.EnsureNumApisixTlsCreated(1))

		backendSvc, backendSvcPort := s.DefaultHTTPBackend()
		route := fmt.Sprintf(`
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: TLSRoute
metadata:
  name: basic-tls-route
spec:
  hostnames: ["%s"]
  rules:
  - backendRefs:
    - name: %s
      port: %d
`, host, backendSvc, backendSvcPort[0])

		assert.Nil(ginkgo.GinkgoT(), s.CreateResourceFromString(route), "creating TLSRoute")
		assert.Nil(ginkgo.GinkgoT(), s.EnsureNumApisixStreamRoutesCreated(1), "Checking number of routes")
		assert.Nil(ginkgo.GinkgoT(), s.EnsureNumApisixUpstreamsCreated(1), "Checking number of upstreams")

		client := s.NewAPISIXClientWithTLSOverTCP(host)
		client.GET("/ip").WithHost(host).Expect().Status(http.StatusOK)
	})
```

So I think I might have missed something in my code. Maybe creating `ApisixRoute` resource only isn't sufficient?

**Comment by tao12345666333** [comment]
Please refer to https://apisix.apache.org/docs/apisix/certificate/

The certificate you create should contain the domain name you requested

**Review by Juneezee — COMMENTED** [review]
I am still clueless about the cause of the error. The certificate is configured with the same domain name as the SNI in `ApisixSsl` resource. Below of the log output of the created `ApisixSsl`:

```go
ssls, err := s.ListApisixSsl()
assert.Nil(ginkgo.GinkgoT(), err)
ginkgo.GinkgoT().Logf("%+v", *ssls[0])
```
```
{ID:2e62297b Snis:[a.test.com] Cert:-----BEGIN CERTIFICATE-----
  MIIDBTCCAe2gAwIBAgIRAMR/vosXRWCr3OSM7O1CUNEwDQYJKoZIhvcNAQELBQAw
  EjEQMA4GA1UEChMHQWNtZSBDbzAeFw0yMjEyMTcxMDE0MjRaFw0yMjEyMTcxMTE0
  MjRaMBIxEDAOBgNVBAoTB0FjbWUgQ28wggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAw
  ggEKAoIBAQCj/dnqgYzYB16sDHO3X+6BgxE5G6v+rSS5sTyza0PfS/VBU6h7xe7H
  Bo9LH7la7bKrHsJOn9gIpIWvQgBYq4kqzBBbUgliSXMefFVtwDOqlDHmNkYMpkgZ
  LFnMrmRg+eYa6cgV1t0eBBwXRZ2dABL9tWk4jmPJA7wR6VIq9WRihnYf/v+0EHga
  j7DqXnaLSr3wppmDFlstuk49GJI2/DrPZB/2ef/xq4qzdQfhfOtgOObYfzGsE8/Q
  lrZxNNpYhglhnyicu+xV67u2oMPO+H9owm+2jd8Uo/w/G4Us5i9PydYqFP4EXQ5g
  IOogWc/iqr22lUmNhCMXjAKvTSy1RM9lAgMBAAGjVjBUMA4GA1UdDwEB/wQEAwIF
  oDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDAYDVR0TAQH/BAIwADAV
  BgNVHREEDjAMggphLnRlc3QuY29tMA0GCSqGSIb3DQEBCwUAA4IBAQCfLKFm+7Te
  R9/zDM8L0ETiWfMHirLRJVdS0I2WUME78WWOKx++DDm+XJKcL++BM/AEeB9mXE7l
  KFXAlX3tbNgGCfX/Z8QPawv3QyK2wma+BvkC6O7lzgkE1Qh51pErKRt6g5AkJoDj
  xFR/E28KBZ9geJ2NYWnZLZwjXijvRDIbEsLYyvcwcCWZJVPe0Wn3REH7cDLiiffG
  62mmeWl8xGoyfCfL6S78ak97gBR8M274sl7pCc7ZyXyoSl0guERjUn2CMXygEL6B
  EEWTNH76sndKX7FP/bezUfZqcAjdb1PttUEYf3lbNoSUCkRFaErPabVwqxaqtMb6
  K5Bg38vVGNWC
  -----END CERTIFICATE-----
   Key:YnwwDKc5vNzo0OU4StTRQbwgCnTZ3dmYiBFm8aGnvTwxKvJJ1WgSomNREpjNtjohs1L21CS1IT5LT8yk+6RKOjybG23TZnTzogGYh4yNfsBU+EK2HktlikqcSus8fYL/3Z0DNw8fEfVXXgOPoAqng+l+nI9WPuyJD9G1gaGg/AdfV9Q5cALw1idqeQzFNorIdiSwzfBi2p5q+NjtJA5T+1tFf21UGee/rTD618wqwrrap8jnAKmbJaxFFdKJs3wYzjoyHhOmjGrTV+b096xYSVLSBvB8eLPFoNxbRwFijXaPWmjHzXm9VqZBMudpet6jyMw5vRPj52T+JL5AdkU3cLkZm8wM8sqvPqRSPwOlC/89/Ply9WK12uj8frnj0Ur/UR4/dBrXTAbAMJYgSseQD8hGcJ3AUkp++OSu2vDl2FGVv0l7hg3TKeJL9bFtwX2+qfIZSYRwsbwinhAc2Rk43KOfT/ghxWseJQA4lFCecref6TPWxoZ6z3nsp1w0+aETfB91YH3rvEQYB8rmbK7mJ+OBEDvU8QlmsMhfBvEQBXd5K0Udf7N7LLq4uuzHZKIygtwNQtU9NYHqUMMUi7Ndk0Pg+sR786Q+sbP0TRlTv746ZmY3p6scL1G8RolW2hJh24Pg6hjWOMShUvO8LmIN36DYLuckgTm2GU60VM3TKFAQP+Q6F+XMwuiU6CU6yxtIwulHycdPasgzW2y3AvM+er+4Iv3ubhcdNOEiMaefCAHaNGwjaCwuOiKuinbjrlgK12vF5isJx3PI0gBXXJ2RlJweY4rnP//5+f8c29vKZYh3DemLpKQWHTO3bkoedF7C/K0HSXtH1SIoH74lTdZeHImPpcDuK5mTHU3fULXlbr6ZqeWEuEWdm2glyQhyachfNkPlN3xbU9qMKPsgkD8cXckcss+/Gk+7ZPeluITc4gbD4TBHhr1If7fAmvGLoLYd3qwc6IXYXtapQg+wuveKmbxRHRQYIP0RaOmpyTCCmQbk9nxQwWfz5A8oRDgEttXEAa9RwnkiuV9roAaG7gXjwLbTLdryRcLXXZsKUocdqJbr0PxCBcdho7Dt7gKtMw3G+54xBXw4MGEKzy95//WPt7faGpQw/GBqhgKXYjE2xyIJb+eeeiRQ1pv8WhdgoQnZ6WZ7zvWWgsl+phZlLbKn/Zwt72zfkyOKzFP9MxktdbIuBZS+xgABdPVxGboF/i3kXH0nB4ElFBeG7HpH/5U1yGamhsEKSTrojLPiMkO3zG/hyjP1GMhtRMAoYjZPKg4zcKPKOlCUdZHNipzaMfFSNi6H/JdMP8FsvKBDIeVzbft4UGaRwpigZbR1Pa2BFfKaLY9cJmYsP8mr0dTxQIAy+ifTQpYs2wfM7oGj5AGrev/9UiHYjQf4rjEFM3NBDckvIL081kMC8K2qyFrcDBILZ8yypHKbujfXwhfmwm6eIuy5tp3DZPRXbhJ4vuMm+FgGv2yIlCIX9mFCDhUFVAclRrSjaARlnHbodzfLeUtRFz7+KjaViYjTZPSc6DY905VRH7APy3JETkN8HOKXYPBlx9zbk4X43QEuKzfV4j2kryDOauZTyeZZnI+/QInK8+38GSkwFQjkf0TmyP4OZWV160FGP8x7qQJNawd6BZk2ZHeq81NckxtD1AYk895j4w2S/gZRpG5KwjKwJ6MwlY9zFY7J7+kT6OIaqrwRBrLCwxC/b859we6dg1bCkbWDOMOAzODyhN85TwcwbKdf4zbDPaYVl+qRJgiH8CkhfvXQsptEtSKTV2G0QuzT6vPymUsmo33t0qioJrssFPNn/r6m5G38rXapwuA1hLmUM3vNTNME1xdyaQlr9KWBlnv4FZ5+vZjQNPl/gksEy31c3wC9Dawu7NqmZKyZ+bhVNJXswqbonkcE5gEcnK/W2+mhmYbyiXoJRi9BCOVch75u1rU8SAG2uhVVWsEsVWm2ZEDaJKCAsC+2Y4AEcEoq9g6830mzJb43WEOwXDCID2nfS2DNPG0YVqh1Ab+6DBo7bvtxX9qivOLSk4avKlS3s3u0Y/+UVRtxwL/GeY1iF6hZ04v7IVfQOLm2tZXtoJ/ClBu0GBSUohauN4TLuuaf83nkqTOYST/g0hTHXUViqThPrv+uOkA74PjtHIQG1phFfc26bs3I0LqS7he5/qSrbOHQTsN4pzJY+HlWSHC1HjqZVLojPNG30eKJOZYO1d+LFQDOEzSbQC31joff8E5MnC2xLT7383+aOu3O8P5+4K4FmSrbJe5JoOhhXDyO6rWD7b1koQw= Status:1 Labels:map[managed-by:apisix-ingress-controller meta_secret_name:server-secret meta_secret_namespace:ingress-apisix-e2e-tests-default-654694026] Client:<nil>}
```

---

I have attached some debugging screenshots below

image

image

The error is a `http.transportReadFromServerError`. I have found a similar issue on GitHub: https://github.com/golang/go/issues/53472

**Comment by tao12345666333** [comment]
@AlinsRan PTAL

**Review by AlinsRan — APPROVED** [review]

File: test/e2e/suite-ingress/suite-ingress-resource/stream.go [hunk 1]
```diff
@@ -214,4 +223,82 @@ spec:
...
+apiVersion: apisix.apache.org/v2
+kind: ApisixRoute
+metadata:
+  name: httpbin-tcp-route
+spec:
+  stream:
+  - name: rule1
+    protocol: TCP
+    match:
+      ingressPort: 9100
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply test/e2e/suite-ingress/suite-ingress-resource/stream.go:253]: 9100 tls is not enabled

**Review by AlinsRan — CHANGES_REQUESTED** [review]

**Comment by Juneezee** [comment]
@AlinsRan Sorry I made a typo :bowing_man: . Please rerun the tests. Thank you

**Comment by tao12345666333** [comment]
Re-run all test cases

File: test/e2e/suite-ingress/suite-ingress-resource/stream.go [hunk 2]
```diff
@@ -214,4 +223,82 @@ spec:
...
+      serviceName: %s
+      servicePort: %d
+`, host, backendSvc, backendSvcPort[0])
+
+		assert.Nil(ginkgo.GinkgoT(), s.CreateVersionedApisixResource(apisixRoute))
+
+		err = s.EnsureNumApisixStreamRoutesCreated(1)
+		assert.Nil(ginkgo.GinkgoT(), err, "Checking number of routes")
+
+		client := s.NewAPISIXClientWithTLSOverTCP(host)
```
Comment (AlinsRan) [inline test/e2e/suite-ingress/suite-ingress-resource/stream.go:265]: I recommend adding the use of other host cases here to ensure that the configured sni is valid.

**Review by AlinsRan — APPROVED** [review]

**Review by AlinsRan — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]

**Pull request merged**